jena-arq/src/main/java/org/apache/jena/sparql/core/TransactionHandlerView.java

```diff
@@ -35,21 +35,41 @@ public TransactionHandlerView(DatasetGraph dsg) {
 
     protected DatasetGraph getDSG() { return dsg; }
 
+    // There may be transactions on this thread to other graphs in the dataset, or a
+    // dataset-level transaction. Test for an existing dataset transaction.
+    // Because this is "same-thread", transactions must be nested ("must" = "should have been"!).
+    //
+    // A fresh TransactionHandlerView is created for every call of getTransactionHandler()
+    // in GraphView and GraphUnionRead.
+    private int dsgTransaction = 0;
+
     @Override
     public void begin() {
-        getDSG().begin(TxnType.READ_PROMOTE);
+        if ( getDSG().isInTransaction() ) {
+            dsgTransaction++;
+        } else {
+            getDSG().begin(TxnType.READ_PROMOTE);
+        }
     }
 
     @Override
     public void abort() {
-        getDSG().abort();
-        getDSG().end();
+        if ( dsgTransaction > 0 ) {
+            dsgTransaction--;
+        } else {
+            getDSG().abort();
+            getDSG().end();
+        }
     }
 
     @Override
     public void commit() {
-        getDSG().commit();
-        getDSG().end();
+        if ( dsgTransaction > 0 ) {
+            dsgTransaction--;
```

Comment:
Is there a potential problem here where the inner transaction calls `abort()` which just decrements the counter but then the outer transaction commits and actually commits the transaction i.e. you get inconsistent states wrt the callers expectations?

Reply:
Good to ask. 

I think it's safe (comment on the declaration of `dsgTransaction`).

Proper nested transactions isn't supported in Jena - only nesting, but same transaction (inner does not begin-commit/abort). Proper nested transactions would need more complex machinery in the base transaction code than exists.

The reported bug happens across two graphs, both from the same dataset, in unrelated code.

A `TransactionHandlerView` is a use-once, same thread object - `GraphView` and `GraphUnionRead` both do a "new" when `getTransactionHandler` is called. 

Being per-transaction, `dsgTransaction` is going to be 0 or 1.  `isInTransaction()` is true if this thread is in the transaction.
Only one `TransactionHandlerView` will see `isInTransaction()` false, then do the real transaction `begin()` that it needs to undo (nesting assumed). As this is same-thread there isn't a concurrency issue of test-and-set.

It would be a problem if the code does non-nested actions on two graphs, same thread; it would be illegal in real nested transactions.

Solution: use `execute` where possible.